core/src/main/java/hudson/model/User.java

```diff
@@ -494,13 +494,46 @@ public synchronized void doSubmitDescription( StaplerRequest req, StaplerRespons
 
     /**
      * Gets the {@link User} object by its id or full name.
+     *
+     * Creates a user on-demand.
+     *
+     * <p>
      * Use {@link #getById} when you know you have an ID.
+     * In this method Jenkins will try to resolve the {@link User} by full name with help of various
+     * {@link hudson.tasks.UserNameResolver}.
+     * This is slow (see JENKINS-23281).
+     *
+     * @deprecated This method is deprecated, because it causes unexpected {@link User} creation
+     *             by API usage code.
+     *             Use {@link #getById} when you know you have an ID.
+     *             Otherwise use {@link #getOrCreate(String)} or {@link #get(String, boolean, Map)}.
      */
+    @Deprecated
     public static @Nonnull User get(String idOrFullName) {
-        return get(idOrFullName,true);
+        return getOrCreate(idOrFullName);
     }
 
     /**
+     * Get the user y ID or Full Name.
+     *
+     * If the user does not exist, creates a new one on-demand.
+     *
+     * <p>
+     * Use {@link #getById} when you know you have an ID.
+     * In this method Jenkins will try to resolve the {@link User} by full name with help of various
+     * {@link hudson.tasks.UserNameResolver}.
+     * This is slow (see JENKINS-23281).
+     *
+     * @param idOrFullName User ID or full name
+     * @return User instance
+     * @since TODO
+     */
+    public static @Nonnull User getOrCreate(@Nonnull String idOrFullName) {
```

Comment:
If we have to introduce a new method taking `idOrFullName`, we should encode the fact that it is doing a `fullName` lookup in the method name itself. Very few callers should ever need to do this. Really just some SCM plugins.